app/server-frontend.js

```diff
@@ -1,15 +1,58 @@
 const express = require('express');
+const basicAuth = require('express-basic-auth');
+const otp = require('node-2fa');
 const path = require('path');
 const config = require('config');
+const fs = require('fs');
 
 const { configureWebSocket } = require('./websocket/configure');
 const { configureLocalTunnel } = require('./local-tunnel/configure');
 
+let OTP_SECRET = false;
+if (fs.existsSync('./.2fa_secret')) {
+  OTP_SECRET = fs.readFileSync('./.2fa_secret', { encoding: 'utf8' });
+}
+
+const basicAuthChallenge = (username, password) => {
+  if (!config.frontend_auth) {
+    return true;
+  }
+
+  const usernameOk = basicAuth.safeCompare(
+    username,
+    config.frontend_auth.username
+  );
+
+  let passwordOk = false;
+  if (config.frontend_auth.mode_otp && OTP_SECRET) {
+    const passwordPlain = basicAuth.safeCompare(password.slice(0, -6), config.frontend_auth.password);
+    const password2fa = Boolean(otp.verifyToken(OTP_SECRET, password.replace(config.frontend_auth.password, '')))
+
+    passwordOk = passwordPlain && password2fa;
+  } else {
+    passwordOk = basicAuth.safeCompare(password, config.frontend_auth.password);
+  }
+
+  return usernameOk & passwordOk;
+};
+
 const runFrontend = async serverLogger => {
   const logger = serverLogger.child({ server: 'frontend' });
   logger.info({ config }, `API ${config.get('mode')} frontend started on`);
 
   const app = express();
+  if (config.frontend_auth) {
+    const baOptions = {
+      authorizer: basicAuthChallenge,
+      challenge: true,
+      realm: 'Binance Trading Bot'
+    };
+    if (config.frontend_auth.mode_otp) {
```

Comment:
I have tried with OTP mode on, but without OTP entered, I still can access the frontend. 

![image](https://user-images.githubusercontent.com/5715919/111893604-07599a80-8a58-11eb-8557-ff484071d45e.png)

![image](https://user-images.githubusercontent.com/5715919/111893624-29ebb380-8a58-11eb-8243-951bd3471218.png)

![image](https://user-images.githubusercontent.com/5715919/111893613-193b3d80-8a58-11eb-93b3-e9797a33a9cb.png)